src/main/java/io/neonbee/data/DataVerticle.java

```diff
@@ -490,20 +547,53 @@ public Future<T> execute(DataQuery query, DataContext context) {
                 // ignore the result of the require data composite future (otherwiseEmpty), the retrieve data method
                 // should decide if it needs to handle success or failure of any of the individual asynchronous results
                 return CompositeFuture.join(Optional.ofNullable(requests).map(Collection::stream).orElse(Stream.empty())
-                        .map(request -> requestResults.computeIfAbsent(request,
-                                mapRequest -> requestData(vertx, request, context.copy())))
-                        .map(Future.class::cast).collect(Collectors.toList())).otherwiseEmpty();
+                        .map(request -> requestResults.computeIfAbsent(request, mapRequest -> {
+                            Future<Object> future = requestData(vertx, request, context.copy());
+                            reportRequestDataMetrics(request, future);
+                            return future;
+                        })).map(Future.class::cast).collect(Collectors.toList())).otherwiseEmpty();
             }).compose(requiredCompositeOrNothing -> {
+                List<Tag> tags = retrieveDataTags();
                 try {
-                    return retrieveData(query, new DataMap(requestResults), context);
+                    Future<T> future = retrieveData(query, new DataMap(requestResults), context);
+                    reportRetrieveDataMetrics(tags, future);
+                    return future;
                 } catch (Exception e) {
+                    dataVerticleMetrics.reportStatusCounter("retrieve_data_counter_" + getAddress(),
```

Comment:
is this actually the right format? shouldn't it be:

```suggestion
                    dataVerticleMetrics.reportStatusCounter("retrieve.data.counter." + getAddress(),
```